apiserver/facades/client/spaces/shims.go

```diff
@@ -99,3 +99,14 @@ func (s *stateShim) SubnetByCIDR(cidr string) (networkingcommon.BackingSubnet, e
 	}
 	return networkingcommon.NewSubnetShim(result), nil
 }
+
+func (s *stateShim) IsControllerModel() (bool, error) {
```

Comment:
You can also call this directly in `NewRenameSpaceModelOp` without getting the model first.

Reply:
I pushed a change. I am not entirely sure what you mean. That's why I added the `stateShim` as a field to the `opFactory`

Reply:
The factory that has both a state and a state indirection as members should be a red flag about the design here.

Consider this simplification (no tests):
```diff
diff --git a/apiserver/facades/client/spaces/opfactory.go b/apiserver/facades/client/spaces/opfactory.go
index 4171267035..66cb3bb292 100644
--- a/apiserver/facades/client/spaces/opfactory.go
+++ b/apiserver/facades/client/spaces/opfactory.go
@@ -20,12 +20,11 @@ type OpFactory interface {
 }
 
 type opFactory struct {
-	st   *state.State
-	shim *stateShim
+	st *state.State
 }
 
-func newOpFactory(st *state.State, shim *stateShim) OpFactory {
-	return &opFactory{st: st, shim: shim}
+func newOpFactory(st *state.State) OpFactory {
+	return &opFactory{st: st}
 }
 
 // NewRemoveSpaceModelOp (OpFactory) returns an operation
@@ -54,9 +53,5 @@ func (f *opFactory) NewRenameSpaceModelOp(fromName, toName string) (state.ModelO
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	isController, err := f.shim.IsControllerModel()
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	return NewRenameSpaceModelOp(isController, controllerSettings, &renameSpaceStateShim{f.st}, space, toName), nil
+	return NewRenameSpaceModelOp(f.st.IsController(), controllerSettings, &renameSpaceStateShim{f.st}, space, toName), nil
 }
diff --git a/apiserver/facades/client/spaces/remove.go b/apiserver/facades/client/spaces/remove.go
index cc069aa125..93295815a2 100644
--- a/apiserver/facades/client/spaces/remove.go
+++ b/apiserver/facades/client/spaces/remove.go
@@ -199,11 +199,8 @@ func (api *API) getConstraintsTagsPerSpace(spaceName string) ([]names.Tag, error
 
 func (api *API) getSpaceControllerSettings(spaceName string) ([]string, error) {
 	var matches []string
-	is, err := api.backing.IsControllerModel()
-	if err != nil {
-		return matches, errors.Trace(err)
-	}
-	if !is {
+
+	if !api.backing.IsController() {
 		return matches, nil
 	}
 
diff --git a/apiserver/facades/client/spaces/shims.go b/apiserver/facades/client/spaces/shims.go
index 844c34d67a..462f9a29d3 100644
--- a/apiserver/facades/client/spaces/shims.go
+++ b/apiserver/facades/client/spaces/shims.go
@@ -19,8 +19,11 @@ func NewStateShim(st *state.State) (*stateShim, error) {
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	return &stateShim{EnvironConfigGetter: stateenvirons.EnvironConfigGetter{State: st, Model: m},
-		State: st, modelTag: m.ModelTag()}, nil
+	return &stateShim{
+		EnvironConfigGetter: stateenvirons.EnvironConfigGetter{State: st, Model: m},
+		State:               st,
+		model:               m,
+	}, nil
 }
 
 // stateShim forwards and adapts state.State methods to Backing
@@ -28,11 +31,11 @@ func NewStateShim(st *state.State) (*stateShim, error) {
 type stateShim struct {
 	stateenvirons.EnvironConfigGetter
 	*state.State
-	modelTag names.ModelTag
+	model *state.Model
 }
 
 func (s *stateShim) ModelTag() names.ModelTag {
-	return s.modelTag
+	return s.model.ModelTag()
 }
 
 func (s *stateShim) AddSpace(name string, providerId network.Id, subnetIds []string, public bool) error {
@@ -51,11 +54,7 @@ func (s *stateShim) SpaceByName(name string) (networkingcommon.BackingSpace, err
 
 // AllEndpointBindings returns all endpoint bindings and maps it to a corresponding common type
 func (s *stateShim) AllEndpointBindings() ([]ApplicationEndpointBindingsShim, error) {
-	model, err := s.State.Model()
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	endpointBindings, err := model.AllEndpointBindings()
+	endpointBindings, err := s.model.AllEndpointBindings()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
@@ -100,17 +99,6 @@ func (s *stateShim) SubnetByCIDR(cidr string) (networkingcommon.BackingSubnet, e
 	return networkingcommon.NewSubnetShim(result), nil
 }
 
-func (s *stateShim) IsControllerModel() (bool, error) {
-	model, err := s.State.Model()
-	if err != nil {
-		return false, errors.Trace(err)
-	}
-	if !model.IsControllerModel() {
-		return false, nil
-	}
-	return true, nil
-}
-
 func (s *stateShim) ConstraintsBySpaceName(spaceName string) ([]Constraints, error) {
 	found, err := s.State.ConstraintsBySpaceName(spaceName)
 	if err != nil {
diff --git a/apiserver/facades/client/spaces/spaces.go b/apiserver/facades/client/spaces/spaces.go
index 1e5d7badcf..c72547123c 100644
--- a/apiserver/facades/client/spaces/spaces.go
+++ b/apiserver/facades/client/spaces/spaces.go
@@ -80,9 +80,6 @@ type Backing interface {
 	// ApplyOperation applies a given ModelOperation to the model.
 	ApplyOperation(state.ModelOperation) error
 
-	// IsControllerModel Returns whether the current model is the controller model
-	IsControllerModel() (bool, error)
-
 	// ControllerConfig Returns the controller config
 	ControllerConfig() (jujucontroller.Config, error)
 
@@ -167,7 +164,7 @@ func NewAPI(st *state.State, res facade.Resources, auth facade.Authorizer) (*API
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	return newAPIWithBacking(stateShim, common.NewBlockChecker(st), state.CallContext(st), res, auth, newOpFactory(st, stateShim))
+	return newAPIWithBacking(stateShim, common.NewBlockChecker(st), state.CallContext(st), res, auth, newOpFactory(st))
 }
 
 // newAPIWithBacking creates a new server-side Spaces API facade with
```

Even so, we don't really have unit tests for the factory, but let's not tackle that here.

Reply:
Thanks that really explains a lot! I just didn't realize that you meant that we already have a method for that in state. My bad!